### ISSUE TITLE
WCF Method Name

### ISSUE THREAD

**Issue by DanielLandryfnf** [microsoft/ApplicationInsights-Home#439]
Hello,

I've added app insights into a wcf service. Ideally I'd like to grab the name of the method so I can run queries against the data coming in and view trends, detailed information, etc. 

I was able to do this by adding the below into a CustomServiceHost.cs. However, while this gave me exactly what I needed we found that this caused severe overhead onto our service as this code runs synchronously.

`var extraTelemetry = new ApplicationInsightsHelper`
`{`
`MethodName = methodName`
`};`

`Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration.Active.TelemetryInitializers.Add(extraTelemetry);`

Are there other alternatives to grabbing the name of the method that won't affect performance besides a custom telemetry initializer? I feel as though this is something that should be supported "out of the box" and perhaps I just haven't found the correct way to do it.

Thanks, Daniel

**Comment by Dmitry-Matveev** [comment]
[Telemetry Initializers](https://docs.microsoft.com/en-us/azure/azure-monitor/app/api-filtering-sampling) are indeed a recommended way to enrich/edit telemetry with the properties you'd like to use.

Telemetry Initializers are executed synchronously to the `Track..()` methods that are called on your behalf by auto-collection mechanism to track Request Telemetry. However, it's a bit of a surprise that this is performing poorly when you add single initializer to the existing list - there are already about 10 default initializers typically, so adding one that changes one variable is not supposed to lead to any perf issues.

One guess, is that method name may not be available for some telemetry types when initializer is executed for them and it may slow things down, so check like `if (telemetry is RequestTelemetry request) {...}` within initializer may help.

Another guess is that retrieving method name can be a heavy procedure in itself. If yes, you may somehow store it at the moment you know it / have it handy without the extraction operations and then use it in the initializer. I could've suggested AsyncLocal but not sure if WCF is good in preserving AsyncLocal across the request execution (if there is managed-unmanaged transition, some weird things may happen with the state, as well as it depends on the WCF pipeline itself). Another way is to store it into HttpContext.Items, but that requires a setting in WCF to enable HttpContext (which will slow down WCF a bit anyway...). Maybe, there is another temp storage for method name that you can use, then it may be a feasible workaround.

**Comment by DanielLandryfnf** [comment]
@Dmitry-Matveev I'm sorry for the delay in response, we had a couple other more pressing changes come in and I had to delay in getting back to this. 

However, I'm happy to say that your guess about how I'm retrieving the method name being a heavy process turned out to be the issue. I found a different, much more efficient method and so far performance seems to be a non-issue (we're waiting on a formal load test to confirm).

**Thanks for your help!**
